### ISSUE TITLE
[feat] an equivalent for `gen_statem`

### ISSUE THREAD

**Issue by bunopnu** [gleam-lang/otp#50]
Hi,

The [gen_statem ](https://www.erlang.org/doc/man/gen_statem) behaviour provides a powerful way to manage stateful processes, especially for development of complex state machines and stateful systems.

It would be great to see something similar for Gleam.

**Comment by lpil** [comment]
Could you go into more detail as the specific requirements for this new process abstraction, including the sorts of problems you would solve using it, and what a typed API might look like? Thank you.

**Comment by bunopnu** [comment]
#### Requirements

I don't think it should be actually same with `gen_statem`, but should be a module that simplifies the process of creating FSMs in Gleam, making FSM implementation more accessible and straightforward for most of the case.

#### Use Cases

Many services, like chat servers and database connections involve maintaining state across multiple interactions. The new equivalent can provide a clean way to manage such states. It can be also used to manage resources like database connections and network sockets for ensuring proper initialization, use, and cleanup.

#### Implementation

I'm currently considering an approach based on the [actor module](https://hexdocs.pm/gleam_otp/gleam/otp/actor.html) to seamlessly integrate the new abstraction within the existing hierarchy. To explore this idea further, I plan to develop a prototype library that demonstrates how something similar to `gen_statem` can be implemented in Gleam.

**Comment by lpil** [comment]
Given Gleam is a typed language how could a library create a finite state machine? They are typically created using types, which the programmer defines.

Assuming there's a good design for a library creating FSMs, why would it be part of `gleam_otp` rather than a stand alone package that isn't coupled to a particular concurrency abstraction?

**Comment by bunopnu** [comment]
> Given Gleam is a typed language how could a library create a finite state machine? They are typically created using types, which the programmer defines.

I don't think that being a typed language is an issue here, as there are already libraries available for typed languages. Checking their implementations may assist us in developing one for Gleam.

> Assuming there's a good design for a library creating FSMs, why would it be part of gleam_otp rather than a stand alone package that isn't coupled to a particular concurrency abstraction?

You are correct; I had considered this repository because it's within Erlang/OTP. Maybe leaving this to the community might be a better choice...

**Comment by lpil** [comment]
Could you share some of those libraries? I was unable to find any for well typed languages, and the ways I know of making FSMs don't have anything that be extracted in to a library.

**Comment by bunopnu** [comment]
I've come across several state machine libraries:

1. [Stateless](https://github.com/dotnet-state-machine/stateless)
2. [Stateless (for Go)](https://github.com/qmuntal/stateless)
3. [ZigFSM](https://github.com/cryptocode/zigfsm)

I'm not entirely certain if these libraries employ any behind-the-scenes magic. Particularly, the ZigFSM appears quite clean and straightforward.

Maybe we can also consider the Elm architecture, as it resembles me a state machine.

**Comment by bunopnu** [comment]
It's my fault since I'm not entirely sure how the type system works in Gleam. I used to believe that functions in Gleam could optionally have their argument and return types annotated, which would somehow enable library users to use their own Event and Message types.

However, I'm currently feeling confused about this. 😅

**Comment by lpil** [comment]
I think a good way to approach this would be to implement some of these protocols and see what sort of code arises when doing it in Gleam. Once we've done a few iterations of this we can figure out if there's any abstractions that can be lifted out of those production tested solutions. Trying to make a solution without specific problems often results in a weaker solution in my opinion.

**Comment by dvic** [comment]
Just sharing a nice example of using gen_statem in Elixir: https://andrealeopardi.com/posts/connection-managers-with-gen-statem

**Comment by Ridderby** [comment]
State machines (like in Erlang gen_statem) differs from gen_servers in that it has a state. In Erlang the state is described by an Atom.

Instead of receiving messages based on (Message) one receives it based on (State, Message). Erlang has a callback framework for this based on the state atom. 

This implies that a message can be treated differently between states and I'd say this is one of the core features in state machines. 

For example: A door that is open (state open) can be Closed  but a Closed door can't. When in state Open it acts on message Close and enter state Closed. When in Closed it ignores message Close and stays in state Closed. In both Open and Closed states it might act on Open but slight different (open a bit more if already open unless opened to it's max angle). 

This can be implemented without an obvious state machine but it will be there anyway hidden in the door logic to make it work as it is supposed to. 

It should be fairly simple to accomplish state machines in gleam by adding the state to the receive switch statement.

I would prefer to have one separate receive switch for each state and that the part that the framework is supposed to add and why I think it is a given to include in Gleam/OTP. I think state machines are vital in a foundation for distributed systems, that it is a sister to Actor/gen_server and equally important. 

There are more to it if anyone up for a deeper discussion but this, I think, is the fundamentals.

**Comment by lpil** [comment]
Thanks for the info @Ridderby !

I'm going to close this as we don't yet have a clear motivating use case for a state machine abstraction that is coupled to a process. If anyone has one please open an issue to discuss 🙏

**Comment by Ridderby** [comment]
I thought that was what I did.  I have a ton of use cases for proper state machines (as in processes instead of the Actor). 

How do you want them described to make sense?

**Comment by lpil** [comment]
You have described state machines in the abstract but not shared a concrete use case that motivates them or why it would need to be process based instead of regular pure code.

**Comment by Ridderby** [comment]
A state machine is fundamentally a reactive (living) entity that communicates with signals (like messages). Many times it has high demands on responsiveness. I would go so far to say that an Actor is a special case of state machine with only one state and what ever reason there is to make the Actor a process, same applies for the state machine. 

In my mind the question is rather the other way around - why would one decide to NOT implement a state machine as a process if there are means to do that.

The use of state machines are very common in protocol handlers, and that is also where the state machine as a process is beneficial. 

Lately I developed a websocket client as a two level state machine, one for the link level and one for the application protocol level. It made the whole thing more easy to understand and build and I could get rid och many corner cases. I could also easy design it in SDL graphs before coding. The framework (Qt/C++ for Web Assembly, single threaded) does not support processes though, but it does have messages and are reactive based both events and Signals and Slots.  The server side in Erlang, has a matching state machine as a gen_statem.

In some other cases like, a markdown parser, I would probably have the state in a class/struct instead so that is a valid alternative in some cases but I think those may not be so easy to standardize.

**Comment by lpil** [comment]
If you wish to propose an addition to Gleam you will need to provide concrete use cases that motivate an addition today. That proposal should detail the problem encountered, what solutions you have tried, and the evidence in favour of the particular design proposed. In this case it would also need to evidence why the state machine library would need to be tightly coupled to processes instead of being pure, as is typical.

We do not accept proposals when there is no real-world problems to motivate them. That would be a solution without a problem.

**Comment by Ridderby** [comment]
I won't spend more time on this, it just doesn't seem to be productive. Sorry for putting my nose in here unasked and thanks for the time I might have taken from more important stuff.

**Comment by lpil** [comment]
Not at all, thank you for your input! Unfortunately it's quite a lot of R&D work as we have to be very conservative with additions

**Comment by schonfinkel** [comment]
I faced the same issue and ended up crafting an [FFI-based lib](https://github.com/byzantine-systems/eparch) with the help of a friend. I'm using `gleam/otp` in one of my projects and I faced a situation where I really needed some typed interface to both `gen_statem` and `gen_event` , I couldn't reinvent the well and I was already familiar with both of these Erlang behaviors.

I really love the fact that Gleam makes it really easy to do Erlang FFIs, but long term I want to investigate if we can generate this layer via some meta-programming (either on Gleam or Erlang's side). There are still a few edges here and there, but I believe we are supporting most of the `gen_statem` / `gen_event` features, long term we should have a way to return `Result`-based errors in case someone tries to use features from an unsupported OTP version, right now I'm assuming you'll be running on the latest OTP.

**Comment by lpil** [comment]
Very cool! Could you share the use case and requirements that resulted in your needing these behaviours please 🙏

**Comment by schonfinkel** [comment]
> Very cool! Could you share the use case and requirements that resulted in your needing these behaviours please 🙏

I'm currently trying to build tooling for infrastructure management on the BEAM (mostly experiments in Erlang / Gleam), it's all I can say for now since it's not ready yet. 

`gen_statem` is more ergonomic than `Actor` / `gen_servers` whenever I need an easy way to [Postpone](https://www.erlang.org/doc/apps/stdlib/gen_statem.html#event-postponing) a message, perform a side-effect operation right upon entering a new state (which is supported by leveraging [state_enter/0](https://www.erlang.org/doc/apps/stdlib/gen_statem.html#t:state_enter/0)) or do some [fancy timeout logic based on the state machine's state instead](https://www.erlang.org/doc/apps/stdlib/gen_statem.html#t:state_timeout/0), like "if a state machine stays in the `AwaitingConnection` state for 1m, trigger a timeout", such a timeout is automatically cancelled if it moves to a different state within the allowed interval (`gen_statem` manages such timeouts automatically for me).

These were the use cases that made me go for a wrapper in the meantime, it could be also Gleam skill issues and one can do all of that with within `Actor`, anyhow, I'm more familiar with Erlang and Gleam enables me to leverage it whenever I need it.

**Comment by lpil** [comment]
I think due to Gleam's types it would be more ergonomic to do that with actor, but it could be that there's a process based state machine API that I have not thought of.
